### PR TITLE
Present fog with canvas instead of SVG

### PR DESCRIPTION
**Pull request by jdbann**
Performance with SVG was very poor. Drawing it with canvas seems better.

Changes the `FogArea` paths to JSON representations rather than the SVG path format.

### CODE REVIEW

File: app/models/fog_area.rb [hunk 1]
```diff
@@ -1,8 +1,6 @@
 class FogArea < ApplicationRecord
-  BASIC_SVG_PATH = /M (?:-?\d+ ){2}(?:L (?:-?\d+ ){2})*Z/.freeze
-
   belongs_to :map
 
   validates :map, presence: true
-  validates :path, presence: true, format: { with: BASIC_SVG_PATH }
+  validates :path, presence: true
```
Comment (jdbann) [inline app/models/fog_area.rb:5]: Didn't think it was worth declaring `path` as a JSON serialized attribute as Rails never needs access to the JSON representation. But then we get the `safe_json` mess in the helper above...

File: db/migrate/20200711183641_convert_fog_path_format_to_array.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    end
+  end
+
+  def down
+    FogArea.find_each do |fog_area|
+      svg_path = JSON.parse(fog_area.path).map.with_index do |segment, index|
+        "#{index.zero? ? 'M' : 'L'} #{segment['x']} #{segment['y']}"
+      end.join(" ") + " Z"
+
+      fog_area.update_attribute(:path, svg_path)
```
[the remark this answers is not in the document]
Comment (jdbann) [reply db/migrate/20200711183641_convert_fog_path_format_to_array.rb:29]: Intentionally skipping validations here

File: db/migrate/20200711183641_convert_fog_path_format_to_array.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        items
+      end
+
+      json_path = segments.map do |segment|
+        next if segment[0].casecmp("z").zero?
+
+        { x: segment[1].to_i, y: segment[2].to_i }
+      end.compact.to_json
+
+      fog_area.update_attribute(:path, json_path)
```
[the remark this answers is not in the document]
Comment (jdbann) [reply db/migrate/20200711183641_convert_fog_path_format_to_array.rb:19]: Intentionally skipping validations here

**Comment by jdbann** [comment]
> I was thinking that it was cool that we were able to do what we were doing without canvas. But I understood the risk of performance issues.

I'm quite sad to be moving to canvas, especially with SVG and Stimulus working so well together. I've been thinking about whether we could use web components to create a version of canvas which uses the DOM...

```html
<table-canvas width="1024" height="768">
  <table-canvas-group composite-operation="source-over">
    <table-canvas-rect width="1024" height="768"/>
  </table-canvas-group>

  <table-canvas-group composite-operation="destination-out">
    <table-canvas-path path="[{"x": 120, "y": "120"},...]"/>
    <table-canvas-path path="[{"x": 120, "y": "120"},...]"/>
    <table-canvas-path path="[{"x": 120, "y": "120"},...]"/>
  </table-canvas-group>
</table-canvas>
```

The `TableCanvasWidth` element can have a `MutationObserver` on its children which triggers a redraw...

> Now that we've introduced it, I do wonder whether it makes sense to move the entire map to the canvas.

Could you expand a bit more on what else you would consider adding to the canvas? Are there any particular parts you think would be easier to handle with canvas instead?

The tokens are layered between the map and the fog, so we would have to either have the map and fog as two separate layers, or add the tokens into the canvas as well. I think that would mean too much work detecting what was clicked on etc.

Right now, the canvas redraws are limited to just changes to the fog or animating the fog in for players.

**Comment by cpytel** [comment]
Yes, I think your absolutely right that that would be the issue with moving everything to the canvas and that would be a bad idea. I wasn't thinking through the layering.

**Pull request merged**